To run this, press "*Runtime*" and press "*Run all*" on a **free** Tesla T4 Google Colab instance!
<div class="align-center">
<a href="https://unsloth.ai/"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
<a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord button.png" width="145"></a>
<a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a></a> Join Discord if you need help + ⭐ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐
</div>

To install Unsloth on your own computer, follow the installation instructions on our Github page [here](https://docs.unsloth.ai/get-started/installing-+-updating).

You will learn how to do [data prep](#Data), how to [train](#Train), how to [run the model](#Inference), & [how to save it](#Save)


### News

**Read our [Gemma 3 blog](https://unsloth.ai/blog/gemma3) for what's new in Unsloth and our [Reasoning blog](https://unsloth.ai/blog/r1-reasoning) on how to train reasoning models.**

Visit our docs for all our [model uploads](https://docs.unsloth.ai/get-started/all-our-models) and [notebooks](https://docs.unsloth.ai/get-started/unsloth-notebooks).


### Installation

In [1]:
%%capture
import os
if "COLAB_" not in "".join(os.environ.keys()):
    !pip install unsloth
else:
    # Do this only in Colab notebooks! Otherwise use pip install unsloth
    !pip install --no-deps bitsandbytes accelerate xformers==0.0.29.post3 peft trl==0.15.2 triton cut_cross_entropy unsloth_zoo
    !pip install sentencepiece protobuf datasets huggingface_hub hf_transfer
    !pip install --no-deps unsloth

### Unsloth

In [2]:
from unsloth import FastLanguageModel  # 从 unsloth 库中导入 FastLanguageModel
import torch  # 导入 PyTorch 库

max_seq_length = 2048  # 最大序列长度，可以设置任意值，内部自动支持 RoPE 扩展
dtype = None  # 数据类型，设置为 None 表示自动检测。Tesla T4 和 V100 使用 Float16，Ampere+ 使用 Bfloat16
load_in_4bit = True  # 是否使用 4bit 量化以减少内存使用，可以设置为 False

# 支持 4bit 预量化模型，这些模型下载速度更快且不会出现内存不足（OOM）
fourbit_models = [
    "unsloth/Meta-Llama-3.1-8B-bnb-4bit",      # Llama-3.1 模型，15万亿训练数据，速度提高 2 倍
    "unsloth/Meta-Llama-3.1-8B-Instruct-bnb-4bit",
    "unsloth/Meta-Llama-3.1-70B-bnb-4bit",
    "unsloth/Meta-Llama-3.1-405B-bnb-4bit",    # 405B 参数模型的 4bit 版本
    "unsloth/Mistral-Nemo-Base-2407-bnb-4bit", # 新版 Mistral 12B，速度提高 2 倍
    "unsloth/Mistral-Nemo-Instruct-2407-bnb-4bit",
    "unsloth/mistral-7b-v0.3-bnb-4bit",        # Mistral v3，速度提高 2 倍
    "unsloth/mistral-7b-instruct-v0.3-bnb-4bit",
    "unsloth/Phi-3.5-mini-instruct",           # Phi-3.5，速度提高 2 倍
    "unsloth/Phi-3-medium-4k-instruct",
    "unsloth/gemma-2-9b-bnb-4bit",
    "unsloth/gemma-2-27b-bnb-4bit",            # Gemma 模型速度提高 2 倍
]  # 更多模型请访问 https://huggingface.co/unsloth

# 加载预训练模型和分词器
model, tokenizer = FastLanguageModel.from_pretrained(
    # 可以从以下模型中选择：
    # "unsloth/Qwen2.5-0.5B", "unsloth/Qwen2.5-1.5B", "unsloth/Qwen2.5-3B"
    # "unsloth/Qwen2.5-14B",  "unsloth/Qwen2.5-32B",  "unsloth/Qwen2.5-72B",
    # 还支持所有 Instruct 版本以及数学、代码相关版本！
    model_name = "unsloth/Qwen2.5-7B",  # 指定要加载的模型名称
    max_seq_length = max_seq_length,    # 设置最大序列长度
    dtype = dtype,                      # 设置数据类型
    load_in_4bit = load_in_4bit,        # 是否启用 4bit 量化
    # token = "hf_...", # 如果使用受限模型（如 Meta-Llama/Llama-2-7b-hf），需要提供访问令牌
)


🦥 Unsloth: Will patch your computer to enable 2x faster free finetuning.
🦥 Unsloth Zoo will now patch everything to make training faster!
==((====))==  Unsloth 2025.3.19: Fast Qwen2 patching. Transformers: 4.50.3.
   \\   /|    Tesla T4. Num GPUs = 1. Max memory: 14.741 GB. Platform: Linux.
O^O/ \_/ \    Torch: 2.6.0+cu124. CUDA: 7.5. CUDA Toolkit: 12.4. Triton: 3.2.0
\        /    Bfloat16 = FALSE. FA [Xformers = 0.0.29.post3. FA2 = False]
 "-____-"     Free license: http://github.com/unslothai/unsloth
Unsloth: Fast downloading is enabled - ignore downloading bars which are red colored!


model.safetensors.index.json:   0%|          | 0.00/106k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/5.00G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/2.54G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/172 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/4.72k [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/2.78M [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/1.67M [00:00<?, ?B/s]

added_tokens.json:   0%|          | 0.00/605 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/617 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

We now add LoRA adapters so we only need to update 1 to 10% of all parameters!

In [3]:
model = FastLanguageModel.get_peft_model(
    model,  # 基础模型
    r = 16,  # LoRA 的秩（rank），可以选择任意大于 0 的数值，建议值为 8、16、32、64、128
    target_modules = ["q_proj", "k_proj", "v_proj", "o_proj",
                      "gate_proj", "up_proj", "down_proj"],  # 指定需要应用 LoRA 的模块
    lora_alpha = 16,  # LoRA 的缩放因子（alpha）
    lora_dropout = 0,  # LoRA 的 dropout 值，支持任意值，但设置为 0 时性能最佳
    bias = "none",     # LoRA 的偏置设置，支持任意值，但设置为 "none" 时性能最佳
    # [新特性] "unsloth" 使用 30% 更少的显存，支持 2 倍更大的批量大小
    use_gradient_checkpointing = "unsloth",  # 启用梯度检查点，支持 True 或 "unsloth"，用于超长上下文
    random_state = 3407,  # 随机种子，用于确保结果可复现
    use_rslora = False,   # 是否使用 Rank Stabilized LoRA（秩稳定 LoRA）
    loftq_config = None,  # 是否使用 LoftQ（量化配置）
)


Unsloth 2025.3.19 patched 28 layers with 28 QKV layers, 28 O layers and 28 MLP layers.


<a name="Data"></a>
### Data Prep
We now use the Alpaca dataset from [yahma](https://huggingface.co/datasets/yahma/alpaca-cleaned), which is a filtered version of 52K of the original [Alpaca dataset](https://crfm.stanford.edu/2023/03/13/alpaca.html). You can replace this code section with your own data prep.

**[NOTE]** To train only on completions (ignoring the user's input) read TRL's docs [here](https://huggingface.co/docs/trl/sft_trainer#train-on-completions-only).

**[NOTE]** Remember to add the **EOS_TOKEN** to the tokenized output!! Otherwise you'll get infinite generations!

If you want to use the `llama-3` template for ShareGPT datasets, try our conversational [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Alpaca.ipynb)

For text completions like novel writing, try this [notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Mistral_(7B)-Text_Completion.ipynb).

In [4]:
# 定义 Alpaca 格式化模板
alpaca_prompt = """以下是描述任务的指令，与提供进一步上下文的输入配对。写一个适当的响应来完成请求。

### Instruction:
{}

### Input:
{}

### Response:
{}"""

EOS_TOKEN = tokenizer.eos_token  # 必须添加 EOS_TOKEN（结束标记）
def formatting_prompts_func(examples):
    instructions = examples["instruction"]  # 从数据集中提取指令
    inputs       = examples["input"]        # 从数据集中提取输入
    outputs      = examples["output"]       # 从数据集中提取输出
    texts = []
    for instruction, input, output in zip(instructions, inputs, outputs):
        # 必须添加 EOS_TOKEN，否则生成过程可能会无限进行
        text = alpaca_prompt.format(instruction, input, output) + EOS_TOKEN
        texts.append(text)
    return { "text" : texts }  # 返回格式化后的文本
pass

# 加载数据集
from datasets import load_dataset
dataset = load_dataset("Thanosx/RegulatoryStandards", split = "train")  # 加载指定数据集
dataset = dataset.map(formatting_prompts_func, batched = True)  # 批量应用格式化函数


README.md:   0%|          | 0.00/63.0 [00:00<?, ?B/s]

apa-secdatasets-1.json:   0%|          | 0.00/6.19M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/4556 [00:00<?, ? examples/s]

Map:   0%|          | 0/4556 [00:00<?, ? examples/s]

<a name="Train"></a>
### Train the model
Now let's use Huggingface TRL's `SFTTrainer`! More docs here: [TRL SFT docs](https://huggingface.co/docs/trl/sft_trainer). We do 60 steps to speed things up, but you can set `num_train_epochs=1` for a full run, and turn off `max_steps=None`. We also support TRL's `DPOTrainer`!

In [5]:
from trl import SFTTrainer  # 从 trl 库中导入 SFTTrainer（微调工具）
from transformers import TrainingArguments  # 从 transformers 库中导入训练参数类
from unsloth import is_bfloat16_supported  # 从 unsloth 库中导入判断是否支持 bfloat16 的函数

trainer = SFTTrainer(
    model = model,  # 要微调的模型
    tokenizer = tokenizer,  # 与模型匹配的分词器
    train_dataset = dataset,  # 训练数据集
    dataset_text_field = "text",  # 数据集中包含文本的字段名称
    max_seq_length = max_seq_length,  # 最大序列长度
    dataset_num_proc = 2,  # 数据集处理时使用的进程数
    packing = False,  # 是否启用数据打包（针对短序列，启用可以加速训练最多 5 倍）
    args = TrainingArguments(
        per_device_train_batch_size = 2,  # 每个设备的训练批量大小
        gradient_accumulation_steps = 4,  # 梯度累计步数（相当于有效批量大小为 2 * 4 = 8）
        warmup_steps = 5,  # 学习率的预热步数
        # num_train_epochs = 1, # 设置完整训练轮数，可启用此参数
        max_steps = 60,  # 最大训练步数
        learning_rate = 2e-4,  # 学习率
        fp16 = not is_bfloat16_supported(),  # 是否启用 FP16（如果不支持 bfloat16，则启用 FP16）
        bf16 = is_bfloat16_supported(),  # 是否启用 bfloat16（判断硬件是否支持）
        logging_steps = 1,  # 日志记录的步数间隔
        optim = "adamw_8bit",  # 优化器类型，使用 8-bit AdamW 优化器以减少显存占用
        weight_decay = 0.01,  # 权重衰减系数
        lr_scheduler_type = "linear",  # 学习率调度器类型，线性衰减
        seed = 3407,  # 随机种子，用于确保结果可复现
        output_dir = "outputs",  # 输出目录，用于保存模型和日志
        report_to = "none",  # 日志报告方式（如 WandB 等），设置为 "none" 表示不使用
    ),
)

Unsloth: Tokenizing ["text"] (num_proc=2):   0%|          | 0/4556 [00:00<?, ? examples/s]

In [6]:
# @title 显示当前显存状态
gpu_stats = torch.cuda.get_device_properties(0)  # 获取 GPU 的属性（设备 0）
start_gpu_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)  # 获取已保留的最大显存，并转换为 GB
max_memory = round(gpu_stats.total_memory / 1024 / 1024 / 1024, 3)  # 获取 GPU 的总显存，并转换为 GB
print(f"GPU = {gpu_stats.name}. 最大显存 = {max_memory} GB.")  # 打印 GPU 名称和总显存
print(f"{start_gpu_memory} GB 的显存已被保留。")  # 打印已保留的显存

GPU = Tesla T4. 最大显存 = 14.741 GB.
7.246 GB 的显存已被保留。


In [ ]:
trainer_stats = trainer.train()

==((====))==  Unsloth - 2x faster free finetuning | Num GPUs used = 1
   \\   /|    Num examples = 4,556 | Num Epochs = 1 | Total steps = 60
O^O/ \_/ \    Batch size per device = 2 | Gradient accumulation steps = 4
\        /    Data Parallel GPUs = 1 | Total batch size (2 x 4 x 1) = 8
 "-____-"     Trainable parameters = 40,370,176/7,000,000,000 (0.58% trained)


In [8]:
# @title 显示最终的显存和时间统计信息
used_memory = round(torch.cuda.max_memory_reserved() / 1024 / 1024 / 1024, 3)  # 获取峰值保留显存，并转换为 GB
used_memory_for_lora = round(used_memory - start_gpu_memory, 3)  # 计算微调过程中使用的显存（LoRA 微调部分）
used_percentage = round(used_memory / max_memory * 100, 3)  # 峰值保留显存占总显存的百分比
lora_percentage = round(used_memory_for_lora / max_memory * 100, 3)  # 微调显存占总显存的百分比

# 打印训练使用的时间（秒）
print(f"{trainer_stats.metrics['train_runtime']} 秒用于训练。")

# 打印训练使用的时间（分钟）
print(f"{round(trainer_stats.metrics['train_runtime']/60, 2)} 分钟用于训练。")

# 打印峰值保留显存（GB）
print(f"峰值保留显存 = {used_memory} GB。")

# 打印微调过程中使用的显存（GB）
print(f"微调过程中使用的峰值显存 = {used_memory_for_lora} GB。")

# 打印峰值保留显存占总显存的百分比
print(f"峰值保留显存占总显存的百分比 = {used_percentage} %。")

# 打印微调显存占总显存的百分比
print(f"微调过程中显存占总显存的百分比 = {lora_percentage} %。")

606.0481 秒用于训练。
10.1 分钟用于训练。
峰值保留显存 = 9.533 GB。
微调过程中使用的峰值显存 = 2.287 GB。
峰值保留显存占总显存的百分比 = 64.67 %。
微调过程中显存占总显存的百分比 = 15.515 %。


<a name="Inference"></a>
### 推理

让我们运行模型吧！你可以更改指令和输入内容——将输出留空即可！

**[新功能] 在免费的 Colab 中体验 Llama-3.1 8b Instruct 的 2 倍速推理，[点击这里](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Unsloth_Studio.ipynb)**

In [ ]:
# alpaca_prompt = 从上方复制的模板
FastLanguageModel.for_inference(model)  # 启用原生的 2 倍速推理模式
inputs = tokenizer(
[
    alpaca_prompt.format(
        "继续斐波那契数列。",  # 指令
        "1, 1, 2, 3, 5, 8",  # 输入
        "",  # 输出 - 留空以进行生成！
    )
], return_tensors="pt").to("cuda")  # 将输入转换为张量并转移到 GPU

outputs = model.generate(**inputs, max_new_tokens=64, use_cache=True)  # 生成最多 64 个新 token
tokenizer.batch_decode(outputs)  # 解码生成的输出

['Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.\n\n### Instruction:\nContinue the fibonnaci sequence.\n\n### Input:\n1, 1, 2, 3, 5, 8\n\n### Response:\n13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6']

 You can also use a `TextStreamer` for continuous inference - so you can see the generation token by token, instead of waiting the whole time!

In [ ]:
# alpaca_prompt = 从上方复制的模板
FastLanguageModel.for_inference(model)  # 启用原生的 2 倍速推理模式
inputs = tokenizer(
[
    alpaca_prompt.format(
        "继续斐波那契数列。",  # 指令
        "1, 1, 2, 3, 5, 8",  # 输入
        "",  # 输出 - 留空以进行生成！
    )
], return_tensors="pt").to("cuda")  # 将输入转换为张量并转移到 GPU

from transformers import TextStreamer  # 导入 TextStreamer
text_streamer = TextStreamer(tokenizer)  # 创建一个文本流对象，用于逐步显示生成内容
_ = model.generate(**inputs, streamer=text_streamer, max_new_tokens=128)  # 使用流式生成，最多生成 128 个新 token

Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
Continue the fibonnaci sequence.

### Input:
1, 1, 2, 3, 5, 8

### Response:
13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 


<a name="Save"></a>
### 保存与加载微调后的模型

要将最终模型保存为 LoRA 适配器，可以使用 Hugging Face 的 `push_to_hub` 方法进行在线保存，或者使用 `save_pretrained` 方法进行本地保存。

**[注意]** 这只会保存 LoRA 适配器，而不会保存完整模型。如果需要保存为 16bit 或 GGUF 格式，请向下滚动查看！

In [ ]:
model.save_pretrained("lora_model")  # 本地保存模型
tokenizer.save_pretrained("lora_model")  # 本地保存分词器
# model.push_to_hub("your_name/lora_model", token="...")  # 在线保存模型到 Hugging Face Hub
# tokenizer.push_to_hub("your_name/lora_model", token="...")  # 在线保存分词器到 Hugging Face Hub

('lora_model/tokenizer_config.json',
 'lora_model/special_tokens_map.json',
 'lora_model/vocab.json',
 'lora_model/merges.txt',
 'lora_model/added_tokens.json',
 'lora_model/tokenizer.json')

现在，如果你想加载我们刚刚保存的 LoRA 适配器进行推理，只需将 `False` 设置为 `True`：

In [ ]:
if False:  # 如果条件为 True，则执行以下代码（当前是 False，代码不会运行）
    from unsloth import FastLanguageModel  # 从 unsloth 库导入 FastLanguageModel
    model, tokenizer = FastLanguageModel.from_pretrained(
        model_name="lora_model",  # 你用于训练的模型名称
        max_seq_length=max_seq_length,  # 最大序列长度
        dtype=dtype,  # 数据类型（例如 fp16、fp32 等）
        load_in_4bit=load_in_4bit,  # 是否以 4bit 格式加载模型（节省显存）
    )
    FastLanguageModel.for_inference(model)  # 启用原生的 2 倍速推理模式

# alpaca_prompt = 你必须从上方复制模板！

inputs = tokenizer(
[
    alpaca_prompt.format(
        "巴黎著名的高塔是什么？",  # 指令
        "",  # 输入
        "",  # 输出 - 留空以生成内容！
    )
], return_tensors="pt").to("cuda")  # 将输入转换为 PyTorch 张量，并移动到 GPU

from transformers import TextStreamer  # 从 transformers 库导入 TextStreamer
text_streamer = TextStreamer(tokenizer)  # 创建一个文本流对象，用于逐步显示生成内容
_ = model.generate(**inputs, streamer=text_streamer, max_new_tokens=128)  # 使用流式生成，最多生成 128 个新 token


Below is an instruction that describes a task, paired with an input that provides further context. Write a response that appropriately completes the request.

### Instruction:
What is a famous tall tower in Paris?

### Input:


### Response:
One of the most famous tall towers in Paris is the Eiffel Tower. It was built by Gustave Eiffel in 1889 and stands at a height of 324 meters (1,063 feet). It is a symbol of Paris and one of the most recognizable structures in the world.<|endoftext|>


你也可以使用 Hugging Face 的 `AutoModelForPeftCausalLM`。仅在未安装 `unsloth` 的情况下使用此方法。不过，这可能会非常慢，因为不支持下载 `4bit` 模型，而使用 Unsloth 的**推理速度快 2 倍**。

In [ ]:
if False:  # 如果条件为 True，则执行以下代码（当前是 False，代码不会运行）
    # 我强烈不推荐使用这种方式 - 如果可能，请使用 Unsloth
    from peft import AutoPeftModelForCausalLM  # 从 peft 库导入 AutoPeftModelForCausalLM
    from transformers import AutoTokenizer  # 从 transformers 库导入 AutoTokenizer
    model = AutoPeftModelForCausalLM.from_pretrained(
        "lora_model",  # 你用于训练的模型名称
        load_in_4bit=load_in_4bit,  # 是否以 4bit 格式加载模型（节省显存）
    )
    tokenizer = AutoTokenizer.from_pretrained("lora_model")  # 加载分词器


### 保存为 float16 格式以供 VLLM 使用

我们支持直接将模型保存为 `float16` 格式。可以选择 `merged_16bit` 保存为 float16 格式，或者选择 `merged_4bit` 保存为 int4 格式。此外，还支持使用 `lora` 适配器作为备用选项。

使用 `push_to_hub_merged` 方法可以将模型上传到你的 Hugging Face 账户！访问 [https://huggingface.co/settings/tokens](https://huggingface.co/settings/tokens) 获取你的个人访问令牌。

In [ ]:
# 合并为 16bit 模型
if False:
    model.save_pretrained_merged("model", tokenizer, save_method="merged_16bit",)
if False:
    model.push_to_hub_merged("hf/model", tokenizer, save_method="merged_16bit", token="")

# 合并为 4bit 模型
if False:
    model.save_pretrained_merged("model", tokenizer, save_method="merged_4bit",)
if False:
    model.push_to_hub_merged("hf/model", tokenizer, save_method="merged_4bit", token="")

# 仅保存 LoRA 适配器
if False:
    model.save_pretrained_merged("model", tokenizer, save_method="lora",)
if False:
    model.push_to_hub_merged("hf/model", tokenizer, save_method="lora", token="")


### GGUF / llama.cpp 转换

现在我们已原生支持将模型保存为 `GGUF` / `llama.cpp` 格式！我们会克隆 `llama.cpp` 并默认保存为 `q8_0` 格式，同时支持其他量化方法，例如 `q4_k_m`。可以使用 `save_pretrained_gguf` 方法进行本地保存，或者使用 `push_to_hub_gguf` 方法上传到 Hugging Face Hub。

---

#### 支持的部分量化方法（完整列表请查看我们的 [Wiki 页面](https://github.com/unslothai/unsloth/wiki#gguf-quantization-options)）：
1. **`q8_0`**：
   - 转换速度快。
   - 资源使用较高，但通常可以接受。
2. **`q4_k_m`**：
   - 推荐使用。
   - 对 `attention.wv` 和 `feed_forward.w2` 张量的一半使用 Q6_K，其余部分使用 Q4_K。
3. **`q5_k_m`**：
   - 推荐使用。
   - 对 `attention.wv` 和 `feed_forward.w2` 张量的一半使用 Q6_K，其余部分使用 Q5_K。

---

#### [**新功能**] 自动导出至 Ollama
如果需要微调并自动导出模型到 Ollama，可以尝试我们的 [Ollama Notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)。

---

### 总结

1. **原生支持 GGUF / llama.cpp**：
   - 使用 `save_pretrained_gguf` 保存模型到本地。
   - 使用 `push_to_hub_gguf` 上传模型到 Hugging Face Hub。

2. **量化方法选择**：
   - `q8_0`：高效转换，资源使用较高。
   - `q4_k_m` 和 `q5_k_m`：推荐量化方法，平衡精度和资源使用。

3. **Ollama 集成**：
   - 提供一个 Colab Notebook，用于微调和自动导出到 Ollama 平台，适合快速部署场景。

In [ ]:
# 保存为 8bit Q8_0 格式
if False:
    model.save_pretrained_gguf("model", tokenizer,)
# 请记得访问 https://huggingface.co/settings/tokens 获取访问令牌！
# 并将 "hf" 替换为你的用户名！
if False:
    model.push_to_hub_gguf("hf/model", tokenizer, token="")

# 保存为 16bit GGUF 格式
if False:
    model.save_pretrained_gguf("model", tokenizer, quantization_method="f16")
if False:
    model.push_to_hub_gguf("hf/model", tokenizer, quantization_method="f16", token="")

# 保存为 q4_k_m GGUF 格式
if False:
    model.save_pretrained_gguf("model", tokenizer, quantization_method="q4_k_m")
if False:
    model.push_to_hub_gguf("hf/model", tokenizer, quantization_method="q4_k_m", token="")

# 保存为多种 GGUF 格式 - 如果需要多种格式，这样速度更快！
if False:
    model.push_to_hub_gguf(
        "hf/model",  # 将 "hf" 替换为你的用户名！
        tokenizer,
        quantization_method=["q4_k_m", "q8_0", "q5_k_m",],  # 指定多种量化方法
        token="",  # 在 https://huggingface.co/settings/tokens 获取访问令牌
    )


Now, use the `model-unsloth.gguf` file or `model-unsloth-Q4_K_M.gguf` file in llama.cpp or a UI based system like Jan or Open WebUI. You can install Jan [here](https://github.com/janhq/jan) and Open WebUI [here](https://github.com/open-webui/open-webui)

And we're done! If you have any questions on Unsloth, we have a [Discord](https://discord.gg/unsloth) channel! If you find any bugs or want to keep updated with the latest LLM stuff, or need help, join projects etc, feel free to join our Discord!

Some other links:
1. Train your own reasoning model - Llama GRPO notebook [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.1_(8B)-GRPO.ipynb)
2. Saving finetunes to Ollama. [Free notebook](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3_(8B)-Ollama.ipynb)
3. Llama 3.2 Vision finetuning - Radiography use case. [Free Colab](https://colab.research.google.com/github/unslothai/notebooks/blob/main/nb/Llama3.2_(11B)-Vision.ipynb)
6. See notebooks for DPO, ORPO, Continued pretraining, conversational finetuning and more on our [documentation](https://docs.unsloth.ai/get-started/unsloth-notebooks)!

<div class="align-center">
  <a href="https://unsloth.ai"><img src="https://github.com/unslothai/unsloth/raw/main/images/unsloth%20new%20logo.png" width="115"></a>
  <a href="https://discord.gg/unsloth"><img src="https://github.com/unslothai/unsloth/raw/main/images/Discord.png" width="145"></a>
  <a href="https://docs.unsloth.ai/"><img src="https://github.com/unslothai/unsloth/blob/main/images/documentation%20green%20button.png?raw=true" width="125"></a>

  Join Discord if you need help + ⭐️ <i>Star us on <a href="https://github.com/unslothai/unsloth">Github</a> </i> ⭐️
</div>
